README.rst
Foris controller data_collect module
====================================
This is a data_collect module for foris-controller

Requirements
============

* foris-controller

Installation
============

	``python setup.py install``


CHANGELOG.rst
1.1 (2018-11-30)
----------------

* setup.py: cleanup + PEP508 updated

1.0 (2018-10-23)
----------------

* moved from orginal foris-controller repository to a separate package


tests/__init__.py
#
# foris-controller-data_collect-module
# Copyright (C) 2018 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#



tests/conftest.py
#
# foris-controller-data_collect-module
# Copyright (C) 2018 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import pytest
import os

# load common fixtures
from foris_controller_testtools.fixtures import (
    ubusd_acl_path, uci_config_default_path, cmdline_script_root,
    controller_modules, extra_module_paths, message_bus, backend
)


@pytest.fixture(scope="session")
def ubusd_acl_path():
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), "ubus-acl")


@pytest.fixture(scope="session")
def uci_config_default_path():
    return os.path.join(
        os.path.dirname(os.path.realpath(__file__)), "uci_configs"
    )


@pytest.fixture(scope="session")
def cmdline_script_root():
    return os.path.join(
        os.path.dirname(os.path.realpath(__file__)), "test_root"
    )


@pytest.fixture(scope="module")
def controller_modules():
    return ["data_collect"]


def pytest_addoption(parser):
    parser.addoption(
        "--backend", action="append",
        default=[],
        help=("Set test backend here. available values = (mock, openwrt)")
    )
    parser.addoption(
        "--debug-output", action="store_true",
        default=False,
        help=("Whether show output of foris-controller cmd")
    )


def pytest_generate_tests(metafunc):
    if 'backend' in metafunc.fixturenames:
        backend = set(metafunc.config.option.backend)
        if not backend:
            backend = ['mock']
        metafunc.parametrize("backend_param", backend, scope='module')


tests/test_data_collect.py
#
# foris-controller
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import pytest
import textwrap

from .conftest import cmdline_script_root
from foris_controller_testtools.fixtures import (
    infrastructure, uci_configs_init, ubusd_test, init_script_result, lock_backend,
    only_backends, FILE_ROOT_PATH
)
from foris_controller_testtools.utils import check_service_result, get_uci_module, FileFaker


@pytest.fixture(
    params=[(200, "free"), (200, "foreign"), (0, "unknown"), (404, "not_found"), (200, "owned")],
    ids=["free", "foreign", "unknown", "not_found", "owned"],
    scope="function"
)
def register_cmd(request, cmdline_script_root):
    status_code, status = request.param

    if not status_code:
        content = """\
            #!/bin/sh
            exit 1
        """
    else:
        content = """\
            #!/bin/sh
            cat <<-EOF
            status: %(status)s
            url: "https://some.page/${2:-en}/data?email=${1}&registration_code=XXXXXXX"
            code: %(code)d
            EOF
        """ % dict(code=status_code, status=status)

    with FileFaker(
        cmdline_script_root, "/usr/share/server-uplink/registered.sh",
        True, textwrap.dedent(content)
    ) as f:
        yield f, status


@pytest.fixture(scope="function")
def registration_code(request):
    content = "0000000B00009CD6"
    with FileFaker(
        FILE_ROOT_PATH, "/usr/share/server-uplink/registration_code",
        False, textwrap.dedent(content)
    ) as f:
        yield f, content


@pytest.mark.parametrize("code", ["cs", "nb_NO"])
def test_get_registered(code, uci_configs_init, infrastructure, ubusd_test):
    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "get_registered",
        "kind": "request",
        "data": {
            "email": "test@test.test",
            "language": code,
        }
    })
    assert "status" in res["data"].keys()


@pytest.mark.only_backends(['openwrt'])
@pytest.mark.parametrize("code", ["cs", "nb_NO"])
def test_get_registered_openwrt(
    cmdline_script_root, code, uci_configs_init, infrastructure, ubusd_test, register_cmd,
    registration_code
):
    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "get_registered",
        "kind": "request",
        "data": {
            "email": "test@test.test",
            "language": code,
        }
    })
    _, status = register_cmd
    assert "status" in res["data"].keys()
    assert res["data"]["status"] == status
    assert status not in ["free", "foreign"] or "url" in res["data"]

def test_get_registered_errors(uci_configs_init, infrastructure, ubusd_test):
    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "get_registered",
        "kind": "request"
    })

    assert "errors" in res
    assert "Incorrect input." in res["errors"][0]["description"]

    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "get_registered",
        "kind": "request",
        "data": {
        }
    })
    assert "errors" in res
    assert "Incorrect input." in res["errors"][0]["description"]


def test_get(uci_configs_init, infrastructure, ubusd_test):
    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "get",
        "kind": "request"
    })
    assert set(res["data"]) == {"agreed", "firewall_status", "ucollect_status"}


def test_set(uci_configs_init, init_script_result, infrastructure, ubusd_test):
    def set_agreed(agreed):
        filters = [("data_collect", "set")]
        old_notifications = infrastructure.get_notifications(filters=filters)
        res = infrastructure.process_message({
            "module": "data_collect",
            "action": "set",
            "kind": "request",
            "data": {
                "agreed": agreed
            }
        })
        assert res == {
            u'action': u'set',
            u'data': {u'result': True},
            u'kind': u'reply',
            u'module': u'data_collect'
        }
        notifications = infrastructure.get_notifications(old_notifications, filters=filters)
        assert notifications[-1] == {
            u"module": u"data_collect",
            u"action": u"set",
            u"kind": u"notification",
            u"data": {
                u"agreed": agreed,
            }
        }
        res = infrastructure.process_message({
            "module": "data_collect",
            "action": "get",
            "kind": "request"
        })
        assert res["data"]["agreed"] is agreed

    set_agreed(True)
    set_agreed(False)
    set_agreed(True)
    set_agreed(False)


@pytest.mark.only_backends(['openwrt'])
def test_set_openwrt(
    uci_configs_init, init_script_result, infrastructure, ubusd_test
):
    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "set",
        "kind": "request",
        "data": {
            "agreed": True
        }
    })
    assert res == {
        u'action': u'set',
        u'data': {u'result': True},
        u'kind': u'reply',
        u'module': u'data_collect'
    }
    check_service_result("ucollect", "enable", clean=False)
    check_service_result("ucollect", "restart")

    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "set",
        "kind": "request",
        "data": {
            "agreed": False
        }
    })
    assert res == {
        u'action': u'set',
        u'data': {u'result': True},
        u'kind': u'reply',
        u'module': u'data_collect'
    }
    check_service_result("ucollect", "disable", clean=False)
    check_service_result("ucollect", "stop")


def test_get_honeypots(infrastructure, ubusd_test):
    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "get_honeypots",
        "kind": "request"
    })
    assert {"minipots", "log_credentials"} == set(res["data"].keys())


def test_set_honeypots(infrastructure, ubusd_test):
    filters = [("data_collect", "set_honeypots")]

    def set_honeypots(result):
        notifications = infrastructure.get_notifications(filters=filters)
        res = infrastructure.process_message({
            "module": "data_collect",
            "action": "set_honeypots",
            "kind": "request",
            "data": {
                "minipots": {
                    "23tcp": result,
                    "2323tcp": result,
                    "80tcp": result,
                    "3128tcp": result,
                    "8123tcp": result,
                    "8080tcp": result,
                },
                "log_credentials": result,
            }
        })
        assert res == {
            u'action': u'set_honeypots',
            u'data': {u'result': True},
            u'kind': u'reply',
            u'module': u'data_collect'
        }
        notifications = infrastructure.get_notifications(notifications, filters=filters)
        assert notifications[-1] == {
            u"module": u"data_collect",
            u"action": u"set_honeypots",
            u"kind": u"notification",
            u"data": {
                "minipots": {
                    "23tcp": result,
                    "2323tcp": result,
                    "80tcp": result,
                    "3128tcp": result,
                    "8123tcp": result,
                    "8080tcp": result,
                },
                "log_credentials": result,
            }
        }
        res = infrastructure.process_message({
            "module": "data_collect",
            "action": "get_honeypots",
            "kind": "request"
        })
        assert res == {
            u"module": u"data_collect",
            u"action": u"get_honeypots",
            u"kind": u"reply",
            u"data": {
                "minipots": {
                    "23tcp": result,
                    "2323tcp": result,
                    "80tcp": result,
                    "3128tcp": result,
                    "8123tcp": result,
                    "8080tcp": result,
                },
                "log_credentials": result,
            }
        }

    set_honeypots(True)
    set_honeypots(False)
    set_honeypots(True)
    set_honeypots(False)


@pytest.mark.only_backends(['openwrt'])
def test_set_honeypots_service_restart(
    uci_configs_init, init_script_result, infrastructure, ubusd_test
):
    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "set_honeypots",
        "kind": "request",
        "data": {
            "minipots": {
                "23tcp": True,
                "2323tcp": False,
                "80tcp": True,
                "3128tcp": False,
                "8123tcp": True,
                "8080tcp": False,
            },
            "log_credentials": False,
        }
    })
    assert res == {
        u'action': u'set_honeypots',
        u'data': {u'result': True},
        u'kind': u'reply',
        u'module': u'data_collect'
    }
    check_service_result("ucollect", "restart", True)


@pytest.mark.only_backends(['openwrt'])
def test_set_agreed_uci(
    uci_configs_init, lock_backend, init_script_result, infrastructure,
    ubusd_test
):
    uci = get_uci_module(lock_backend)

    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "set",
        "kind": "request",
        "data": {
            "agreed": True
        }
    })
    assert res == {
        u'action': u'set',
        u'data': {u'result': True},
        u'kind': u'reply',
        u'module': u'data_collect'
    }
    with uci.UciBackend() as backend:
        data = backend.read()

    assert uci.parse_bool(uci.get_option_named(data, "foris", "eula", "agreed_collect", "0"))

    res = infrastructure.process_message({
        "module": "data_collect",
        "action": "set",
        "kind": "request",
        "data": {
            "agreed": False
        }
    })
    assert res == {
        u'action': u'set',
        u'data': {u'result': True},
        u'kind': u'reply',
        u'module': u'data_collect'
    }

    with uci.UciBackend() as backend:
        data = backend.read()

    assert not uci.parse_bool(uci.get_option_named(data, "foris", "eula", "agreed_collect", "0"))


foris_controller_data_collect_module/__init__.py
#
# foris-controller-data_collect-module
# Copyright (C) 2018 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

__version__ = "1.1"


foris_controller_modules/data_collect/handlers/openwrt.py
#
# foris-controller
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import logging

from foris_controller.handler_base import BaseOpenwrtHandler
from foris_controller.utils import logger_wrapper

from foris_controller_backends.data_collect import (
    RegisteredCmds, DataCollectUci, SendingFiles,
)

from .. import Handler

logger = logging.getLogger(__name__)


class OpenwrtDataCollectHandler(Handler, BaseOpenwrtHandler):

    sending_files = SendingFiles()
    registered_cmds = RegisteredCmds()
    uci = DataCollectUci()

    @logger_wrapper(logger)
    def get_registered(self, email, language):
        """ Tries to obtain info whether the user was registered

        :param email: email which will be used during the server query
        :type email: str
        :param language: language which will be used during the server query
        :type language: str
        :returns: result
        :rtype: dict
        """
        return OpenwrtDataCollectHandler.registered_cmds.get_registered(email, language)

    @logger_wrapper(logger)
    def get_agreed(self):
        """ Get information whether the user agreed with data collect
        :returns: True if user agreed, False otherwise
        :rtype: boolean
        """
        return self.uci.get_agreed()

    @logger_wrapper(logger)
    def set_agreed(self, agreed):
        """ Mock setting information whether the user agreed with data collect
        :param agreed: user agreed with data collect (True/False)
        :type agreed: boolean
        :returns: True
        :rtype: boolean
        """
        return self.uci.set_agreed(agreed)

    @logger_wrapper(logger)
    def get_honeypots(self):
        """ Get configuration of the honeypots
        :returns: {"minipots": {...}, "log_credentials": True/False}
        :rtype: dict
        """
        return self.uci.get_honeypots()

    @logger_wrapper(logger)
    def set_honeypots(self, honepot_settings):
        """ Set configuration of the honeypots
        :param honepot_settings: {"minipots": {...}, "log_credentials": True/False}
        :type honepot_settings: dict

        :returns: True
        :rtype: boolean
        """
        return self.uci.set_honeypots(honepot_settings)

    @logger_wrapper(logger)
    def get_sending_info(self):
        """ Obtains info whether the router is sending data to our servers

        :returns: result
        :rtype: dict
        """
        return self.sending_files.get_sending_info()


foris_controller_modules/data_collect/handlers/mock.py
#
# foris-controller
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import logging
import random

from foris_controller.handler_base import BaseMockHandler
from foris_controller.utils import logger_wrapper

from .. import Handler

logger = logging.getLogger(__name__)


class MockDataCollectHandler(Handler, BaseMockHandler):
    agreed = False
    log_credentials = False
    minipots = {
        "23tcp": False,
        "2323tcp": False,
        "80tcp": False,
        "3128tcp": False,
        "8123tcp": False,
        "8080tcp": False,
    }

    @logger_wrapper(logger)
    def get_registered(self, email, language):
        """ Mocks registration info

        :param email: email which was used during the registration
        :type email: str
        :param language: language which will be used in the server query (iso2)
        :type language: str

        :returns: Mocked result
        :rtype: dict
        """
        registration_code = "%016X" % random.randrange(0x10000000000000000)
        return random.choice([
            {
                "status": "free",
                "url": "https://some.page/%s/data?email=%s&registration_code=%s" %
                (language, email, registration_code),
                "registration_number": registration_code,
            },
            {
                "status": "foreign",
                "url": "https://some.page/%s/data?email=%s&registration_code=%s" %
                (language, email, registration_code),
                "registration_number": registration_code,
            },
            {
                "status": "unknown",
            },
            {
                "status": "not_found",
            },
            {
                "status": "owned",
            },
        ])

    @logger_wrapper(logger)
    def get_agreed(self):
        """ Mock getting information whether the user agreed with data collect
        :returns: True if user agreed, False otherwise
        :rtype: boolean
        """
        return self.agreed

    @logger_wrapper(logger)
    def get_sending_info(self):
        """ Returns fake sending status

        :returns: Mocked result
        :rtype: dict
        """
        choices = ["online", "offline", "unknown"]
        return {
            "firewall_status": {"state": random.choice(choices), "last_check": 1501857960},
            "ucollect_status": {"state": random.choice(choices), "last_check": 1501857970},
        }


    @logger_wrapper(logger)
    def set_agreed(self, agreed):
        """ Mock setting information whether the user agreed with data collect
        :returns: True
        :rtype: boolean
        """
        self.agreed = agreed
        return True

    @logger_wrapper(logger)
    def get_honeypots(self):
        """ Mock getting configuration of the honeypots
        :returns: {"minipots": {...}, "log_credentials": True/False}
        :rtype: dict
        """
        return {
            "minipots": self.minipots,
            "log_credentials": self.log_credentials,
        }

    @logger_wrapper(logger)
    def set_honeypots(self, honepot_settings):
        """ Mock setting configuration of the honeypots
        :param honepot_settings: {"minipots": {...}, "log_credentials": True/False}
        :type honepot_settings: dict

        :returns: True
        :rtype: boolean
        """
        self.log_credentials = honepot_settings["log_credentials"]
        self.minipots = honepot_settings["minipots"]
        return True
